### DISCUSSION TITLE
Text recognition & Search

### DISCUSSION THREAD

**Discussion by but-whythough** [ente-io/ente#547] [Enhancements]
being able to process individual photos or bulk process them to make text searchable would be a great addition

**Comment by harrrrshit** [comment]
Yes I fully agree with this, this is a much needed feature. Sometimes you just need a screenshot in which something was written.

**Comment by vishnukvmd** [comment]
Sidenote: https://store.google.com/intl/en_uk/ideas/articles/pixel-screenshots/

**Reply by brog-io** [reply]
Would love some feature for screenshots specifically

**Comment by brog-io** [comment]
Also, being able to copy text from those images

**Comment by tbeest** [comment]
OCR would make it go from decent to amazing for me.  
I've been hoping to find a good third-party gallery app to replace Samsung's.

**Comment by stelioskat** [comment]
This is for me the most missed feature that I used almost daily in apple photos.  Is there any plan on implementing it?

**Comment by martynhare** [comment]
Text recognition now appears to be there as well as searching by text using machine learning features. If you turn on local indexes on mobile, it then downloads the required models, including a text analysis model which seems to work extremely well, at least in my experience. 

The missing part which I think users really want (myself included) is using built-in native on-device APIs for real-time text analysis when users view photos. On iOS for example, adding a Live Text view will give folks that 1:1 parity with what Apple Photos can do by rapidly OCRing the text when a user attempts to highlight it. This would be completely independent of the index ente creates with the models it uses. The Android equivalent would be to use MLKit which returns where in an image the highlightable text is (the bounding boxes) along with the text itself, allowing it to be hghlightable.

As a private, on-device interim workaround for each platform:

GNU/Linux distros: Screenshot then use Tesseract or a frontend which leverages it (such as gImageReader).
Windows 11 SAC: Ensure OneDrive, telemetry and improvement features are all off, then use Snipping Tool.
Windows 10 SAC/LTSC or Windows 11 IoT LTSC: MS PowerToys Text Extractor. Make sure you install OCR packs.
macOS Monterey or newer: Screenshot and copy Live Text from macOS preview app (with iCloud Drive off).
iOS/iPadOS 15 or newer: Screenshot and copy Live Text from preview. If iCloud Photos is off, nothing is sent to Apple.

I hope this helps everyone involved and props to the ente development team for making everything open source and easily self-hostable. Between ente, bitwarden, syncthing and nextcloud, it’s now very possible to self-host with full end to end encryption without needing to intentionally expose anything to the public Internet (in any form) which is super cool.

**Comment by brog-io** [comment]
OCR is now available in the latest release 🥳

**Comment by devkamiki** [comment]
Is it just me or it's impossible to search for the text in a picture? 
If I type "cat", it shows pictures that contain cats, not pictures that contain the word "cat".
Immich however could let the user decide if they are searching context or OCR.
Currently the OCR in ente app is only for extracting text from a certain picture. I would love to know if it's possible to implement the feature ❤️

**Reply by stelioskat** [reply]
It actually does work in search. It might need to finish processing in the background. However, results are not always perfect.

**Reply by devkamiki** [reply]
Ohhh I see!
It's probably because I searched for CJK characters.
for Latin phrases it works pretty well

**Reply by brog-io** [reply]
Btw, currently the actually OCR isn't being indexed and saved. It's per image. So the current search is just the ML that detects the word. So you might not get great results

**Reply by devkamiki** [reply]
Yes, it would be really nice if the OCR results could be indexed, to improve the search accuracy. Currently, I get thousands of relevant results in "magic" if I searched for a CJK string (and the most relevant ones do not show up in the front rows), compared to less than 200 with a latin string.

**Reply by ramenIgirl** [reply]
I second that, I searched for "shipping code" to try to find every screenshot I took of those (those words are obviously in every screenshot) and it showed me instead Instagram stories and other text images with no words related to that
